cdap-docs/developer-manual/source/metadata/field-lineage.rst

```diff
@@ -0,0 +1,195 @@
+
+    @Override
+    public void prepareRun(BatchSourceContext context) throws Exception {
+      if (config.getSchema() != null && config.getSchema().getFields() != null) {
+        List<Schema.Field> fields = config.getSchema().getFields();
+        // Make sure the schema and fields are non null
+        FieldOperation operation = new FieldReadOperation("Read", "Read from files",
+                                                          EndPoint.of(context.getNamespace(), config.referenceName),
+                                                          fields.stream().map(Schema.Field::getName)
+                                                            .collect(Collectors.toList()));
+        context.record(Collections.singletonList(operation));
+      }
+    }
+
+Concatenate plugin concatenates the fields as represented by the plugin config and record the
+``FieldTransformOperation`` in its prepareRun method::
+
+    @Override
+    public void prepareRun(StageSubmitterContext context) throws Exception {
+      FieldOperation operation = new FieldTransformOperation("Concatenate", "Concatenated fields",
+                                                             Arrays.asList(config.fieldToConcatenate1,
+                                                                           config.FieldToConcatenate2),
+                                                             config.newFieldName);
+      context.record(Collections.singletonList(operation));
+    }
+
+Similarly Address Normalizer plugin can record the ``FieldTransformOperation`` representing address normalization::
+
+    @Override
+    public void prepareRun(StageSubmitterContext context) throws Exception {
+      FieldOperation operation = new FieldTransformOperation("Normalize", "Normalized field",
+                                                             Collections.singletonList(config.fieldToNormalize),
+                                                             config.FieldToConcatenate2);
+      context.record(Collections.singletonList(operation));
+    }
+
+Finally, Table sink can record ``FieldWriteOperation`` representing writing fields to dataset::
+
+    @Override
+    public void prepareRun(BatchSinkContext context) throws Exception {
+      if (schema.getFields() != null) {
+        FieldOperation operation = new FieldWriteOperation("Write", "Wrote to CDAP Table",
+                                                           EndPoint.of(context.getNamespace(), "passenger"),
+                                                           schema.getFields().stream().map(Schema.Field::getName)
+                                                             .collect(Collectors.toList()));
+        context.record(Collections.singletonList(operation));
+      }
+    }
+
+Note that individual plugin can record multiple operations and all operations recorded by single plugin **must** have
```

Comment:
Note: Individual plugins can record multiple operations, and all operations recorded by single plugin must have unique names.


>>>>> Question: what happens if the operation does not have a unique name? It's not recorded? Error--does it show in logs?

Reply:
updated per suggestion. If operation does not have unique name then lineage does not get recorded and error is shown in the log.